File: src/cc/redpen/intellij/RedPenProvider.java
package cc.redpen.intellij;

import cc.redpen.ConfigurationLoader;
import cc.redpen.RedPen;
import cc.redpen.RedPenException;
import cc.redpen.config.Configuration;
import cc.redpen.parser.DocumentParser;
import cc.redpen.util.LanguageDetector;
import com.google.common.collect.ImmutableMap;
import com.intellij.psi.PsiFile;

import java.util.LinkedHashMap;
import java.util.Map;

public class RedPenProvider {
  private static RedPenProvider instance;
  private Map<String, Configuration> initialConfigs = new LinkedHashMap<>();
  private Map<String, Configuration> configs = new LinkedHashMap<>();
  private String configKey = "en";
  private boolean autodetect = true;

  Map<String, DocumentParser> parsers = ImmutableMap.of(
    "PLAIN_TEXT", DocumentParser.PLAIN,
    "Markdown", DocumentParser.MARKDOWN,
    "AsciiDoc", DocumentParser.ASCIIDOC
  );

  public static RedPenProvider getInstance() {
    if (instance == null) instance = new RedPenProvider();
    return instance;
  }

  private RedPenProvider() {
    loadConfig("redpen-conf.xml");
    loadConfig("redpen-conf-ja.xml");
    loadConfig("redpen-conf-ja-hankaku.xml");
    loadConfig("redpen-conf-ja-zenkaku2.xml");
    reset();
  }

  public void reset() {
    initialConfigs.forEach((k,v) -> configs.put(k, v.clone()));
  }

  /** For tests */
  RedPenProvider(Map<String, Configuration> configs) {
    this.configs = configs;
    this.initialConfigs = new LinkedHashMap<>(configs);
  }

  private void loadConfig(String fileName) {
    try {
      Configuration configuration = new ConfigurationLoader().loadFromResource("/" + fileName);
      initialConfigs.put(configuration.getKey(), configuration);
    }
    catch (RedPenException e) {
      throw new RuntimeException(e);
    }
  }

  public void addConfig(Configuration config) {
    initialConfigs.put(config.getKey(), config.clone());
    configs.put(config.getKey(), config.clone());
  }

  public RedPen getRedPen() {
    try {
      return new RedPen(configs.get(configKey));
    }
    catch (RedPenException e) {
      throw new RuntimeException(e);
    }
  }

  public RedPen getRedPenFor(String text) {
    if (autodetect) configKey = new LanguageDetector().detectLanguage(text);
    return getRedPen();
  }

  public DocumentParser getParser(PsiFile file) {
    return parsers.get(file.getFileType().getName());
  }

  public Map<String, Configuration> getInitialConfigs() {
    return initialConfigs;
  }

  public Configuration getInitialConfig(String key) {
    return initialConfigs.get(key);
  }

  public Map<String, Configuration> getConfigs() {
    return configs;
  }

  public Configuration getConfig(String key) {
    return configs.get(key);
  }

  public Configuration getActiveConfig() {
    return configs.get(configKey);
  }

  public void setActiveConfig(Configuration config) {
    configKey = config.getKey();
  }

  public void setAutodetect(boolean autodetect) {
    this.autodetect = autodetect;
  }

  public boolean isAutodetect() {
    return autodetect;
  }
}

File: src/cc/redpen/intellij/RedPenProvider.kt
package cc.redpen.intellij

import cc.redpen.RedPen
import cc.redpen.config.Configuration
import cc.redpen.config.ConfigurationExporter
import cc.redpen.config.ConfigurationLoader
import cc.redpen.parser.DocumentParser
import cc.redpen.parser.DocumentParser.*
import cc.redpen.util.LanguageDetector
import com.intellij.openapi.components.SettingsSavingComponent
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.Project.DIRECTORY_STORE_FOLDER
import com.intellij.openapi.util.io.FileUtil
import com.intellij.psi.PsiFile
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.*

open class RedPenProvider : SettingsSavingComponent {
    val project: Project
    var configDir: File

    open var initialConfigs : MutableMap<String, Configuration> = LinkedHashMap()
    open var configs : MutableMap<String, Configuration> = LinkedHashMap()
    internal val configLastModifiedTimes: MutableMap<String, Long> = LinkedHashMap()

    private var configKey = "en"
    internal var configKeysByFile = Properties()

    companion object {
        val parsers: Map<String, DocumentParser> = mapOf(
                "PLAIN_TEXT" to PLAIN,
                "Markdown" to MARKDOWN,
                "MultiMarkdown" to MARKDOWN,
                "AsciiDoc" to ASCIIDOC,
                "Properties" to PROPERTIES)

        val defaultConfigKeys = LinkedHashSet(Configuration.getDefaultConfigKeys())

        fun forProject(project: Project) = project.getComponent(RedPenProvider::class.java)!!
    }

    internal constructor(project: Project) {
        this.project = project
        this.configDir = File(project.basePath + '/' + DIRECTORY_STORE_FOLDER, "redpen")
        availableConfigKeys().forEach { loadConfig(it) }
        loadConfigKeysByFile()
    }

    /** For tests  */
    internal constructor(project: Project, configs: MutableMap<String, Configuration>) {
        this.project = project
        this.configDir = File(System.getProperty("java.io.tmpdir"))
        this.configs = configs.map { it.key to it.value.clone() }.toMap(LinkedHashMap())
        this.initialConfigs = configs.map { it.key to it.value.clone() }.toMap(LinkedHashMap())
    }

    internal fun loadConfig(key: String) {
        val fileName = key + ".xml"
        val loader = ConfigurationLoader()
        try {
            val file = File(configDir, fileName)

            val initialConfig = if (key in defaultConfigKeys) createInitialConfig(key) else loader.load(file)
            initialConfigs[key] = initialConfig

            if (key in defaultConfigKeys && file.exists()) {
                val config = loader.load(file)
                configs[key] = config
            } else {
                configs[key] = initialConfig.clone()
            }

            configLastModifiedTimes[key] = file.lastModified()
        }
        catch (e: Exception) {
            LoggerFactory.getLogger(javaClass).warn("Failed to load " + fileName, e)
        }
    }

    private fun createInitialConfig(key: String) = Configuration.builder(key).addAvailableValidatorConfigs().build()

    internal fun loadConfigKeysByFile() {
        val file = File(configDir, "files.xml")
        if (file.exists()) FileInputStream(file).use { configKeysByFile.loadFromXML(it) }
    }

    override fun save() {
        configDir.mkdirs()
        configs.values.forEach { c ->
            val file = File(configDir, c.key + ".xml")

            if (file.lastModified() > configLastModifiedTimes[c.key] as Long) loadConfig(c.key)
            else if (c.key in defaultConfigKeys && c == initialConfigs[c.key]) file.delete()
            else {
                FileOutputStream(file).use { out -> ConfigurationExporter().export(c, out) }
                configLastModifiedTimes[c.key] = file.lastModified()
            }
        }

        val file = File(configDir, "files.xml")
        if (configKeysByFile.isEmpty) file.delete()
        else FileOutputStream(file).use { out -> configKeysByFile.storeToXML(out, null) }

        if (configDir.list().isEmpty()) configDir.delete()
    }

    internal fun availableConfigKeys() = if (!configDir.exists()) defaultConfigKeys
                                         else defaultConfigKeys + configDir.list().filter { it != "files.xml" && it.endsWith(".xml") }.map { it.replace(".xml", "") }

    infix operator fun plusAssign(config: Configuration) {
        initialConfigs[config.key] = config.clone()
        configs[config.key] = config.clone()
        StatusWidget.forProject(project).rebuild()
    }

    open fun getRedPen(): RedPen = RedPen(configs[configKey])

    open fun getRedPenFor(file: PsiFile): RedPen {
        configKey = getConfigKeyFor(file)
        return getRedPen()
    }

    open fun getConfigKeyFor(file: PsiFile) = configKeysByFile.getProperty(relativePath(file)) ?: LanguageDetector().detectLanguage(file.text)

    open fun getParser(file: PsiFile): DocumentParser? {
        return parsers[file.fileType.name]
    }

    open var activeConfig: Configuration
        get() = configs[configKey]!!
        set(config) {
            configKey = config.key
        }

    open fun setConfig(file: PsiFile, config: Configuration) {
        activeConfig = config
        configKeysByFile[relativePath(file)] = config.key
    }

    internal fun relativePath(file: PsiFile) = FileUtil.getRelativePath(project.basePath!!, file.virtualFile.path, File.separatorChar)
}
